doc/fips/SONiC-OpenSSL-FIPS-140-3.md

```diff
@@ -0,0 +1,144 @@
+
+
+## Abbreviation
+
+| Abbreviation | Description                                  |
+| ------------ | -------------------------------------------- |
+| CAVP         | Cryptographic Algorithm Validation Program   |
+| CST          | Cryptographic and Security Test              |
+| CMVP         | Cryptographic Module Validation Program      |
+| FIPS         | Federal Information Processing Standard      |
+
+## Requirement
+SONiC only uses cryptographic modules validated by FIPS 140-3, Make SONiC compliant with FIPS 140-3.
+
+## The cryptographic modules in SONiC
+
+| Module   | Use Scenarios | Description                                                |
+| -------------------- | --------------- | -------------------------------------------- |
+| OpenSSL              | Python, OpenSSH | Cyptography and SSL/TLS ToolKit              |
+| Kerberos             | OpenSSH         | Kerboros contains builtin crypto module      |
+| Golang               | sonic-restapi   | Golang contains builtin crypto module         |
+| Libgcrypto           | GPG             | A general purpose cryptographic library originally based on code from GnuPG |
+| Kernel Crypto        | --              | Linux crypto kernel module |
+
+## Scopes:
+In Scopes:
+1. OpenSSL
+2. Kerberos
+3. Golang
+
+Out of Scopes:
+1. Linux Kernel
+2. Libgcrypt
+
+
+## OpenSSL FIPS 140-3
+
+![FIPS Overview](images/fips-overview.png)
+
+### OpenSSL Engine
+OpenSSL supports engine cryptographic modules in the form of engine objects, and provides a reference-counted mechanism to allow them to be dynamically loaded in and out of the running application. An engine object can implement one or all cryptographic algorithms.
+
+### SymCrypt OpenSSL Engine
+The [SymCrypt engine for OpenSSL (SCOSSL)](https://github.com/microsoft/SymCrypt-OpenSSL) allows the use of OpenSSL with [SymCrypt](https://github.com/microsoft/SymCrypt) as the provider for core cryptographic operations. It leverages the OpenSSL engine interface to override the cryptographic implementations in OpenSSL's libcrypto. The primary motivation for this is to support FIPS certification, as OpenSSL 1.1.1 does not have a FIPS-certified cryptographic module. Microsoft will submit the FIPS 140-3 reports for SymCrypt to CMVP.
+
+The SymCrypt Engine is one of the implementation to support FIPS, The [wolfSSL engine](https://github.com/wolfSSL/wolfEngine) is another option.
+
+### OpenSSL configuration for SymCrypt Engine
+
+/usr/lib/ssl/openssl-fips.cnf 
+```
+openssl_conf = openssl_init
+[ openssl_init ]
+engines = engine_section
+
+[ engine_section ]
+symcrypt = symcrypt_section
+
+[ symcrypt_section ]
+engine_id = symcrypt
+dynamic_path = /usr/lib/x86_64-linux-gnu/libsymcryptengine.so
+default_algorithms = ALL
+```
+
+### OpenSSL configuration enhancement
+When fips=1 is set in /proc/cmdline, the OpenSSL default config file is changed to "/usr/lib/ssl/openssl-fips.cnf", otherwise, the config file "/usr/lib/ssl/openssl-fips.cnf" is used.
```

Comment:
typo in the default configuration file name (when fips=0)

Reply:
@eddyk-smart , thanks for you comment, I will change the config file, the none fips config (/usr/lib/ssl/openssl.cnf) will be used if fips=0.

Reply:
Fixed.